### DISCUSSION TITLE
Obtain directives within ExecutionStrategy execute function

### DISCUSSION THREAD

**Discussion by Meemaw** [graphql-java/graphql-java#3674] [General]
Hi,

I'm trying to obtain specific directives of all fields in a custom ExecutionStrategy before I execute the query. This would be used to deduce if a query is cacheable, and if we should even consider checking the cache.

Directive looks like this:

```graphql
directive @cacheControl(
  maxAge: Int!
  staleWhileRevalidate: Int!
) on FIELD_DEFINITION | OBJECT | INTERFACE | UNION
```

I tried a few things in the `execute` function using the provided `parameters: ExecutionStrategyParameters`.

1. Reading directives from the field ~ these always seem to be empty

```kotlin
 parameters.fields.subFieldsList.forEach { mergedField ->
        mergedField.fields.forEach {
            println(it.directives)
        }
    }
```

2. Reading directives manually ~ this works for top level field, but having some issues with nested fields. Also it looks wrong, given that the `parentType` is tied to the execution and is not on field leve.

```kotlin
val parentType =
parameters.executionStepInfo.unwrappedNonNullType as GraphQLObjectType
parameters.fields.subFieldsList.forEach { mergedField ->
  mergedField.fields.forEach {
      val fieldDef =
          getFieldDef(executionContext.graphQLSchema, parentType, it)
  
      println("${it.name}: ${CacheControlPolicy.fromDirectiveContainer(fieldDef)}")
  }
}
```

Any help would be appreciated.

**Comment by bbakerman** [comment]
Check out `graphql.analysis.QueryTraverser` - it follows all the graphql "field merging rules" and presents a "visitor" callback API.

```
        QueryTraverser queryTraversal = QueryTraverser.newQueryTraverser()
                .schema(schema)
                .document(document)
                .variables(variables)
                .build()
                
                
queryTraversal.visitPreOrder(visitor)
                
```

Have a read of https://www.graphql-java.com/blog/deep-dive-merged-fields to learn why you cant just navigate AST elements

**Reply by Meemaw** [reply]
Hey, thanks for the pointer. Before I go down this route, how expensive would you say this is? Is this roughly linear ~ O(N) in regards to number of selected fields in the query.

**Reply by bbakerman** [reply]
I suspect O(N) - you can also choose depth first versus breath first
